// zilswap-sdk/lib/constants.js
"use strict";
var _a, _b, _c, _d;
Object.defineProperty(exports, "__esModule", { value: true });
var util_1 = require("@zilliqa-js/util");
var Network;
(function (Network) {
    Network["MainNet"] = "MainNet";
    Network["TestNet"] = "TestNet";
})(Network = exports.Network || (exports.Network = {}));
exports.APIS = (_a = {},
    _a[Network.MainNet] = 'https://api.zilliqa.com',
    _a[Network.TestNet] = 'https://dev-api.zilliqa.com',
    _a);
exports.WSS = (_b = {},
    _b[Network.MainNet] = 'wss://api-ws.zilliqa.com',
    _b[Network.TestNet] = 'wss://dev-ws.zilliqa.com',
    _b);
exports.CONTRACTS = (_c = {},
    _c[Network.MainNet] = 'zil1hgg7k77vpgpwj3av7q7vv5dl4uvunmqqjzpv2w',
    _c[Network.TestNet] = 'zil1rf3dm8yykryffr94rlrxfws58earfxzu5lw792',
    _c);
exports.CHAIN_VERSIONS = (_d = {},
    _d[Network.MainNet] = util_1.bytes.pack(1, 1),
    _d[Network.TestNet] = util_1.bytes.pack(333, 1),
    _d);
exports.BASIS = 10000;
exports.ZIL_HASH = '0x0000000000000000000000000000000000000000';
